crates/storage/libmdbx-rs/src/txn_manager.rs

```diff
@@ -0,0 +1,316 @@
+    const READ_TRANSACTIONS_CHECK_INTERVAL: Duration = Duration::from_millis(500);
+
+    impl TxnManager {
+        pub(crate) fn with_max_read_transaction_duration(
+            mut self,
+            duration: Duration,
+        ) -> TxnManager {
+            let read_transactions = Arc::new(ReadTransactions::new(duration));
+            read_transactions.clone().start_monitor();
+            self.read_transactions = Some(read_transactions);
+
+            self
+        }
+
+        pub(crate) fn add_active_read_transaction(&self, ptr: *mut ffi::MDBX_txn) {
+            if let Some(read_transactions) = &self.read_transactions {
+                read_transactions.add_active(ptr);
+            }
+        }
+
+        pub(crate) fn remove_active_read_transaction(
+            &self,
+            ptr: *mut ffi::MDBX_txn,
+        ) -> Option<(usize, Instant)> {
+            self.read_transactions.as_ref()?.remove_active(ptr)
+        }
+
+        pub(crate) fn remove_aborted_read_transaction(
+            &self,
+            ptr: *mut ffi::MDBX_txn,
+        ) -> Option<usize> {
+            self.read_transactions.as_ref()?.remove_aborted(ptr)
+        }
+    }
+
+    #[derive(Debug, Default)]
+    pub(crate) struct ReadTransactions {
+        /// Maximum duration that a read transaction can be open until the
+        /// [ReadTransactions::start_monitor] aborts it.
+        max_duration: Duration,
+        /// List of currently active read transactions.
+        ///
+        /// We store `usize` instead of a raw pointer as a key, because pointers are not
+        /// comparable. The time of transaction opening is stored as a value.
+        active: DashMap<usize, Instant>,
+        /// List of read transactions aborted by the [ReadTransactions::start_monitor].
+        /// We keep them until user tries to abort the transaction, so we're able to report a nice
+        /// [Error::ReadTransactionAborted] error.
+        ///
+        /// We store `usize` instead of a raw pointer, because pointers are not comparable.
+        aborted: DashSet<usize>,
+    }
+
+    impl ReadTransactions {
+        pub(super) fn new(max_duration: Duration) -> Self {
+            Self { max_duration, ..Default::default() }
+        }
+
+        pub(crate) fn add_active(&self, ptr: *mut ffi::MDBX_txn) {
+            let _ = self.active.insert(ptr as usize, Instant::now());
+        }
+
+        pub(crate) fn remove_active(&self, ptr: *mut ffi::MDBX_txn) -> Option<(usize, Instant)> {
+            self.active.remove(&(ptr as usize))
+        }
+
+        pub(crate) fn add_aborted(&self, ptr: *mut ffi::MDBX_txn) {
+            self.aborted.insert(ptr as usize);
+        }
+
+        pub(crate) fn remove_aborted(&self, ptr: *mut ffi::MDBX_txn) -> Option<usize> {
+            self.aborted.remove(&(ptr as usize))
+        }
+
+        pub(super) fn start_monitor(self: Arc<Self>) {
+            std::thread::spawn(move || {
+                let mut aborted_active = Vec::new();
```

Comment:
should this be moved inside the loop?

Reply:
less allocations on growing the vec capacity, so i'd say no